providers/common/kdfs/krb5kdf.c

```diff
@@ -0,0 +1,407 @@
+        remainder = tmp;
+    }
+    /* resulting a is the GCD, LCM(a,b) = |a*b|/GCD(a,b) */
+    lcm = blocksize * constant_len / gcd;
+
+    /* now spread out the bits */
+    memset(block, 0, blocksize);
+
+    /* last to first to be able to bring carry forward */
+    carry = 0;
+    for (l = lcm - 1; l >= 0; l--) {
+        unsigned int rotbits, rshift, rbyte;
+
+        /* destination byte in block is l % N */
+        b = l % blocksize;
+        /* Our virtual s buffer is R = L/K long (K = constant_len) */
+        /* So we rotate backwards from R-1 to 0 (none) rotations */
+        rotbits = 13 * (l / constant_len);
+        /* find the byte on s where rotbits falls onto */
+        rbyte = l - (rotbits / 8);
+        /* calculate how much shift on that byte */
+        rshift = rotbits & 0x07;
+        /* rbyte % constant_len gives us the unrotated byte in the
+         * constant buffer, get also the previous byte then
+         * appropriately shift them to get the rotated byte we need */
+        tmp = (constant[(rbyte-1) % constant_len] << (8 - rshift)
+               | constant[rbyte % constant_len] >> rshift)
+              & 0xff;
+        /* add with carry to any value placed by previous passes */
+        tmp += carry + block[b];
+        block[b] = tmp & 0xff;
+        /* save any carry that may be left */
+        carry = tmp >> 8;
+    }
+
+    /* if any carry is left at the end, add it through the number */
+    for (b = blocksize - 1; b >= 0 && carry != 0; b--) {
+        carry += block[b];
+        block[b] = carry & 0xff;
+        carry >>= 8;
+    }
+}
+
+static int cipher_init(EVP_CIPHER_CTX *ctx,
+                       const EVP_CIPHER *cipher, ENGINE *engine,
+                       const unsigned char *key, size_t key_len)
+{
+    int klen, ret;
+
+    ret = EVP_EncryptInit_ex(ctx, cipher, engine, key, NULL);
+    if (!ret)
+        goto out;
+    /* set the key len for the odd variable key len cipher */
+    klen = EVP_CIPHER_CTX_key_length(ctx);
+    if (key_len != (size_t)klen) {
+        ret = EVP_CIPHER_CTX_set_key_length(ctx, key_len);
+        if (!ret)
+            goto out;
+    }
+    /* we never want padding, either the length requested is a multiple of
+     * the cipher block size or we are passed a cipher that can cope with
+     * partial blocks via techniques like cipher text stealing */
+    ret = EVP_CIPHER_CTX_set_padding(ctx, 0);
+    if (!ret)
+        goto out;
+
+out:
+    return ret;
+}
+
+static int KRB5KDF(const EVP_CIPHER *cipher, ENGINE *engine,
+                   const unsigned char *key, size_t key_len,
+                   const unsigned char *constant, size_t constant_len,
+                   unsigned char *okey, size_t okey_len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    unsigned char block[EVP_MAX_BLOCK_LENGTH * 2];
+    unsigned char *plainblock, *cipherblock;
+    size_t blocksize;
+    size_t cipherlen;
+    size_t osize;
+    int ret;
+
+    if (key_len != okey_len) {
```

Comment:
I don't think this check functions correctly with 3des.  The 3DES buffer I pass in is 21 in length, which causes this to fail.

Reply:
(3des has key_len 24, but calls will pass in buffers 21 in length).

Reply:
Oh yes, 3des is quite special ... If you see I have some custom logic in there just for it.
Specifically the 3des algorithm internally does use a 21 bytes buffer, however that buffer gets "fixed up" and stretched back to 24 bytes. so the 21 bytes part is an internal detail.

If you are trying to pass a 21 bytes buffer your are using the interface incorrectly and I am glad this check blocked your from continuing, you just proved the usefulness of that check.

In MIT code you should not use krb5_c_random_to_key() on the output of the KDF but either k5_rand2key_direct() or simply fill in a keyblock directly.
The output of this KDF is a full DES3 key.

Reply:
Note: If it is particularly hard to cope with this, I guess I could have a check that accepts both 21 and 24 bytes output buffers for 3des, and then fixes-up the key or not depending on the size of the buffer. However this adds more complexity and more places to gunfoot yourself so I would like to do that only if really unavoidably necessary.

Reply:
Well, it comes about because we're doing DR and DK as separate passes... for other enctypes it doesn't matter, but for this one it does.  I'll see what I can do.

Reply:
The immediate problem I'm having is that SSKDF does not do the 3DES rand-to-key, which makes it odd to do it here.  Conditionalizing whether we do it or not based on which KDF is in use is a bit odd to me.

Reply:
That's annoying, I'll make it conditional so we can see how that makes the code, then I'll let @paulidale or @levitte decide what is better I guess.

Reply:
Okay, that causes my branch to work as-is.